src/test/kotlin/org/rustSlowTests/lang/resolve/CargoProjectResolveTest.kt

```diff
@@ -545,4 +547,189 @@ class CargoProjectResolveTest : RsWithToolchainTestBase() {
+                    fn main() {
+                        common_dep::bar();
+                    }              //^
+                """)
+            }
+        }
+        dir("common_dep") {
+            toml("Cargo.toml", """
+                [package]
+                name = "common_dep"
+                version = "0.1.0"
+                authors = []
+
+                [dependencies]
+
+                [features]
+                foo = []
+                bar = []
+            """)
+
+            dir("src") {
+                rust("lib.rs", """
+                    #[cfg(feature = "foo")]
+                    pub fn foo() {}
+                    #[cfg(feature = "bar")]
+                    pub fn bar() {}
+                """)
+            }
+        }
+    }.run {
+        val prj = create(project, cargoProjectDirectory)
+        project.testCargoProjects.attachCargoProject(cargoProjectDirectory.pathAsPath.resolve("project_1/Cargo.toml"))
+        project.testCargoProjects.attachCargoProject(cargoProjectDirectory.pathAsPath.resolve("project_2/Cargo.toml"))
+        prj.checkReferenceIsResolved<RsPath>("project_1/src/main.rs")
+        prj.checkReferenceIsResolved<RsPath>("project_2/src/main.rs")
+    }
+
+    @MinRustcVersion("1.41.0") // In this version Cargo starts providing `[dev-dependencies]` info
+    fun `test cyclic dev deps`() = buildProject {
+        toml("Cargo.toml", """
+            [package]
+            name = "hello"
+            version = "0.1.0"
+            build = "build.rs"
+
+            [dev-dependencies]
+            foo = { path = "./foo" }
+        """)
+        dir("tests") {
+            rust("main.rs", """
+                extern crate foo;
+                fn main() {
+                    foo::bar();
+                }     // ^
+            """)
+        }
+        dir("src") {
+            rust("lib.rs", """
+                pub fn bar() {}
+                #[test]
+                fn test() {
+                    extern crate foo;
+                    foo::bar();
+                }     // ^
+            """)
+        }
+        dir("foo") {
+            toml("Cargo.toml", """
+                [package]
+                name = "foo"
+                version = "1.0.0"
+
+                [dependencies]
+                hello = { path = "../" }
+            """)
+            dir("src") {
+                rust("lib.rs", """
+                    extern crate hello;
+                    pub use hello::bar;
+                """)
+            }
+        }
+    }.run {
+        CrateGraphTestmarks.cyclicDevDependency.checkHit {
+            checkReferenceIsResolved<RsPath>("tests/main.rs")
+            checkReferenceIsResolved<RsPath>("src/lib.rs")
+        }
+    }
+
+    @MinRustcVersion("1.41.0") // In this version Cargo starts providing `[build-dependencies]` info
+    fun `test build dependency is resolved in 'build rs' and not resolved in 'main rs'`() = buildProject {
```

Comment:
Let's add similar tests when a common dependency is not resolved from `build.rs`